migration_challenge_keras_image/util/nb.py

```diff
@@ -30,10 +30,11 @@ def upload_data(out):
             f"Uploading files from {local_path} to {s3_uri} in the background...\n"
         )
         t0 = time.time()
-        proc = subprocess.Popen(
-            ["aws", "s3", "sync", "--quiet", "--delete", local_path, s3_uri],
+        print(f"Uploading files from {local_path} to {s3_uri} in the background...\n")
+        proc = subprocess.Popen("cd data && ( find train -mindepth 1 -maxdepth 1 -type d -print0 | xargs -n1 -0 -P10 -I {} aws s3 cp --recursive --quiet {}/ s3://sagemaker-ap-southeast-2-705060800786/mnist/{}/ )",
             stdout=subprocess.PIPE,
             stderr=subprocess.STDOUT,
+            shell=True 
         )
```

Comment:
So I see [some claims](https://stackoverflow.com/questions/3172470/actual-meaning-of-shell-true-in-subprocess) that `shell=True` may be an antipattern - and typically only used where needing to resolve some variables in the current environment in the child process. Is there some reason to add it here?

Also you have a hard-coded bucket name in there which we need to fix.

What's the performance impact of using this complex batch `cp` command instead of `sync`? Could we just get away with doing a straight `cp --recursive` for the whole data folder? Or does the performance advantage basically all come from not uploading unnecessary files like the .npy?

Reply:
Updating with my observations now I've had chance to play around:

- Previous function with new (blank) S3 folder: 3m45s, 3m21s (tested twice)
- Sub `sync` for `cp --recursive` and point to a new S3 folder: 3m23s
- `cp --recursive` against a prefix where the data has already been uploaded: 3m24s
- Previous function against a prefix where the data has already been uploaded: 3m21s

...So it doesn't seem like subbing `sync` for `cp` gives a major boost from my end - it's the parallelizing via `xargs` that yields the speedup.

I also don't think this is picking up the `test` folders at all?

Reply:
Partially addressed in 784e8457d0fd008d8d074575975a939c2aae1e18 - happy to discuss further!